File: app/schemas/checkout.py
"""Checkout + invoice + payment schemas."""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.invoice import (
    Invoice,
    InvoiceLine,
    InvoiceStatus,
    Payment,
    PaymentMode,
)


class CheckoutLine(BaseModel):
    """One cart line in the finalize request. The cashier submits the
    barcodes + quantities; the server is the source of truth on price."""

    model_config = ConfigDict(extra="forbid")

    barcode: str = Field(min_length=1, max_length=64)
    quantity: int = Field(gt=0, le=100_000)


class PaymentInput(BaseModel):
    """One payment line. One or more per invoice (D-59, R-40)."""

    model_config = ConfigDict(extra="forbid")

    mode: PaymentMode
    amount: Decimal = Field(gt=Decimal("0"), max_digits=12, decimal_places=2)


class CheckoutFinalizeRequest(BaseModel):
    """The body of POST /checkout/finalize.

    The client must also send an `Idempotency-Key` header — the service
    layer enforces it; the route reads it from the request.
    """

    model_config = ConfigDict(extra="forbid")

    lines: list[CheckoutLine] = Field(min_length=1, max_length=200)
    payments: list[PaymentInput] = Field(min_length=1, max_length=10)
    note: str | None = Field(default=None, max_length=200)
    # Superadmin-only (D-65): names the target shop. Owner/cashier must
    # omit this — their own shop_id is used implicitly.
    shop_id: int | None = Field(default=None)

    @model_validator(mode="after")
    def _payments_nonempty_amount(self) -> CheckoutFinalizeRequest:
        total = sum((p.amount for p in self.payments), Decimal("0"))
        if total <= 0:
            raise ValueError("payments must sum to a positive amount")
        return self


class InvoiceLinePublic(BaseModel):
    """One line item on a finalized invoice.

    ``product_brand`` and ``product_size_label`` are a snapshot captured
    at sale time (issue #38, D-v3-4). Renaming a product later does
    NOT change this line's display. For rows created before the
    snapshot columns shipped, the API resolves them via a live join —
    see ``app/services/_line_snapshots.py`` — so the wire shape is
    always a present string, never ``null``.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    quantity: int
    unit_price: Decimal
    line_total: Decimal
    product_brand: str
    product_size_label: str


class PaymentPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    mode: PaymentMode
    amount: Decimal


class InvoicePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    shop_id: int
    cashier_user_id: int
    cashier_name: str | None
    invoice_number: int
    status: InvoiceStatus
    total_amount: Decimal
    note: str | None
    finalized_at: datetime
    business_date: date
    eod_signed_off: bool
    lines: list[InvoiceLinePublic]
    payments: list[PaymentPublic]


class CheckoutFinalizeResponse(BaseModel):
    invoice: InvoicePublic
    is_replay: bool  # True when an existing Idempotency-Key was matched


class CartValidationLine(BaseModel):
    barcode: str = Field(min_length=1, max_length=64)
    requested_quantity: int = Field(gt=0)
    available_quantity: int = Field(ge=0)
    accepted_quantity: int = Field(ge=0)
    adjusted: bool


class CartValidationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lines: list[CheckoutLine] = Field(min_length=1, max_length=200)
    shop_id: int | None = None


class CartValidationResponse(BaseModel):
    lines: list[CartValidationLine]


class InvoiceEditRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lines: list[CheckoutLine] = Field(min_length=1, max_length=200)
    payments: list[PaymentInput] = Field(min_length=1, max_length=10)
    note: str | None = Field(default=None, max_length=200)


class InvoiceListResponse(BaseModel):
    invoices: list[InvoicePublic]


# Re-exports.
__all__ = [
    "CartValidationLine",
    "CartValidationRequest",
    "CartValidationResponse",
    "CheckoutFinalizeRequest",
    "CheckoutFinalizeResponse",
    "CheckoutLine",
    "Invoice",
    "InvoiceEditRequest",
    "InvoiceLine",
    "InvoiceLinePublic",
    "InvoiceListResponse",
    "InvoicePublic",
    "InvoiceStatus",
    "Payment",
    "PaymentInput",
    "PaymentMode",
    "PaymentPublic",
]

File: frontend/src/api/checkout.ts
// Checkout API helpers — kept thin so the cart state and the wire format
// stay separate. The page component owns the cart; these helpers just hit
// the backend.

import { api, withShopId } from "./client";

export type PaymentMode = "cash" | "upi" | "card" | "other";

export interface CheckoutLine {
  barcode: string;
  quantity: number;
}

export interface PaymentInput {
  mode: PaymentMode;
  amount: string;
}

export interface InvoiceLinePublic {
  id: number;
  product_id: number;
  quantity: number;
  unit_price: string;
  line_total: string;
  // Issue #38 — snapshot of the product's brand + size, captured at
  // sale time. Always present on the wire (the API falls back to a
  // live Product join for pre-migration rows).
  product_brand: string;
  product_size_label: string;
}

export interface PaymentPublic {
  id: number;
  mode: PaymentMode;
  amount: string;
}

export interface InvoicePublic {
  id: number;
  shop_id: number;
  cashier_user_id: number;
  cashier_name: string | null;
  invoice_number: number;
  status: "finalized" | "voided" | "reversal" | "pending_void";
  total_amount: string;
  note: string | null;
  finalized_at: string;
  business_date: string;
  eod_signed_off: boolean;
  lines: InvoiceLinePublic[];
  payments: PaymentPublic[];
}

export interface CheckoutFinalizeResponse {
  invoice: InvoicePublic;
  is_replay: boolean;
}

export function finalizeCheckout(
  payload: { lines: CheckoutLine[]; payments: PaymentInput[]; note?: string },
  idempotencyKey: string,
  shopId?: number | null
): Promise<CheckoutFinalizeResponse> {
  return api<CheckoutFinalizeResponse>("/checkout/finalize", {
    method: "POST",
    json: withShopId(payload, shopId),
    idempotencyKey,
  });
}

export function getInvoice(invoiceId: number): Promise<InvoicePublic> {
  return api<InvoicePublic>(`/invoices/${invoiceId}`);
}

export interface CartValidationLine {
  barcode: string;
  requested_quantity: number;
  available_quantity: number;
  accepted_quantity: number;
  adjusted: boolean;
}

export function validateCheckoutCart(
  lines: CheckoutLine[],
  shopId?: number | null
): Promise<{ lines: CartValidationLine[] }> {
  return api<{ lines: CartValidationLine[] }>("/checkout/validate", {
    method: "POST",
    json: withShopId({ lines }, shopId),
  });
}

export function listInvoices(opts: {
  source: "current" | "past";
  shopId?: number | null;
  dateFrom?: string;
  dateTo?: string;
  cashier?: number;
  paymentMode?: PaymentMode;
  status?: InvoicePublic["status"];
}): Promise<{ invoices: InvoicePublic[] }> {
  const params = new URLSearchParams();
  params.set("source", opts.source);
  if (opts.dateFrom) params.set("date_from", opts.dateFrom);
  if (opts.dateTo) params.set("date_to", opts.dateTo);
  if (opts.cashier) params.set("cashier", String(opts.cashier));
  if (opts.paymentMode) params.set("payment_mode", opts.paymentMode);
  if (opts.status) params.set("status", opts.status);
  if (opts.shopId != null) params.set("shop_id", String(opts.shopId));
  return api<{ invoices: InvoicePublic[] }>(`/invoices?${params.toString()}`);
}

export function editInvoice(
  invoiceId: number,
  payload: { lines: CheckoutLine[]; payments: PaymentInput[]; note?: string }
): Promise<InvoicePublic> {
  return api<InvoicePublic>(`/invoices/${invoiceId}`, { method: "PATCH", json: payload });
}

// Browser-native <a href> doesn't send headers, so a plain link can't
// carry the Bearer token — we fetch the PDF as a blob (via `api()`, so
// auth-header injection and error-detail parsing come for free) and
// hand the caller an object URL to trigger the download with.
export async function downloadInvoicePdf(invoiceId: number): Promise<Blob> {
  return api<Blob>(`/invoices/${invoiceId}/pdf`, { responseType: "blob" });
}
